codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/QueryShapeSerVisitor.java

```diff
@@ -0,0 +1,233 @@
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.model.traits.EnumTrait;
+import software.amazon.smithy.model.traits.TimestampFormatTrait;
+import software.amazon.smithy.model.traits.TimestampFormatTrait.Format;
+import software.amazon.smithy.model.traits.XmlFlattenedTrait;
+import software.amazon.smithy.model.traits.XmlNameTrait;
+import software.amazon.smithy.utils.FunctionalUtils;
+
+/**
+ * Visitor to generate serialization functions for shapes in AWS Query protocol
+ * document bodies.
+ *
+ * This class handles function body generation for all types expected by the
+ * {@code DocumentShapeSerVisitor}. No other shape type serialization is overwritten.
+ *
+ * Timestamps are serialized to {@link Format}.DATE_TIME by default.
+ */
+final class QueryShapeSerVisitor extends DocumentShapeSerVisitor {
+    private static final Format DEFAULT_TIMESTAMP_FORMAT = Format.DATE_TIME;
+    private static final Logger LOGGER = Logger.getLogger(JsonShapeSerVisitor.class.getName());
+
+    private final Predicate<MemberShape> memberFilter;
+
+    public QueryShapeSerVisitor(GenerationContext context) {
+        this(context, FunctionalUtils.alwaysTrue());
+    }
+
+    public QueryShapeSerVisitor(GenerationContext context, Predicate<MemberShape> memberFilter) {
+        super(context);
+        this.memberFilter = memberFilter;
+    }
+
+    private DocumentMemberSerVisitor getMemberSerVisitor(MemberShape member, String source, String dest) {
+        // Get the timestamp format to be used, defaulting to epoch seconds.
+        Format format = member.getMemberTrait(getContext().getModel(), TimestampFormatTrait.class)
+                .map(TimestampFormatTrait::getFormat)
+                .orElse(DEFAULT_TIMESTAMP_FORMAT);
+        return new DocumentMemberSerVisitor(getContext(), source, dest, format);
+    }
+
+    @Override
+    protected Map<String, String> getAdditionalSerArguments() {
+        return Collections.singletonMap("value", "query.Value");
+    }
+
+    @Override
+    protected void serializeCollection(GenerationContext context, CollectionShape shape) {
+        GoWriter writer = context.getWriter();
+        MemberShape member = shape.getMember();
+        Shape target = context.getModel().expectShape(member.getTarget());
+
+        // If the list is empty, exit early to avoid extra effort.
+        writer.write("if len(v) == 0 { return nil }");
+
+        writer.write("array := value.Array($S)", getSerializedLocationName(member, "member"));
+        writer.write("");
+
+        writer.openBlock("for i := range v {", "}", () -> {
+            // Null values should be omitted for query.
+            if (!target.hasTrait(EnumTrait.class)) {
+                writer.openBlock("if vv := v[i]; vv == nil {", "}", () -> {
+                    writer.write("continue");
+                });
+            }
+
```

Comment:
this implies that query does not support sparse collections, correct?

Reply:
Right. There's more debate going on as to whether they should be supported at all, even where it is technically accepted. At any rate this can be easily changed if need be, though I'm not sure how we would serialize a null for query.